src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -182,48 +184,77 @@ private void ThrowIfDisposed()
             }
         }
 
-        private static void Replenish(DefaultPartitionedRateLimiter<TResource, TKey> limiter)
+        private async Task TimerLoop()
         {
-            lock (limiter.Lock)
+            lock (Lock)
             {
-                if (limiter._disposed)
+                if (_disposed)
                 {
                     return;
                 }
 
                 // If the cache has been invalidated we need to recreate it
-                if (limiter._cacheInvalid)
+                if (_cacheInvalid)
                 {
-                    limiter._cachedLimiters.Clear();
-                    bool cacheStillInvalid = false;
-                    foreach (KeyValuePair<TKey, Lazy<RateLimiter>> kvp in limiter._limiters)
+                    _cachedLimiters.Clear();
+                    _cachedLimiters = _limiters.ToList();
```

Comment:
Why call `_cachedLimiters.Clear()` if we're reassigning right afterwards? We should still copy the elements manually to avoid the allocation.